move linux packages out of all-packages

xaverdh opened the pull request:
###### Motivation for this change
see https://github.com/NixOS/nixpkgs/issues/122606


###### Current State
~~This is currently somewhat broken, in that e.g. `linuxPackages` fails to eval for reasons I don't understand (somehow related to makeExtensible). But~~ I wanted to get some code out there, so people can give some feedback.

todo:
  * ~~think about `recurseIntoAttrs`, the way its currently used is probably broken~~ should be fixed, hope I did that correctly
  * improve documentation?

###### Notes on what changed
This moves the kernels and their package sets into `linuxKernel.{kernels,packages}` in `linux-kernels.nix`. It also moves a few related functions and attributes there.
The `linuxKernel.kernels` set is composed of various subsets (rpi kernels, rt kernels, etc.) which can be passed to the new function `linuxKernel.latestPackageWithModules`, to select from them.
It does strictly only contain the versioned kernels themselves. Any unversioned aliases and such are added on top (in `all-packages.nix`). This is because these are considered user facing, e.g. most users will want to "follow" a kernel like `linux_latest` (or use the new function).
I kept the versioned `linuxPackages_*` and `linux_*` aliases for backwards compat., but I think they should either be considered deprecated in favor `linuxKernel.packages.*` and/or generated from these automatically (but I don't know how to that yet; help is very much appreciated here!).
~~I also dropped the aliases for the versioned kernels, these can be obtained from `linuxKernel.kernels` now. In contrast to the package sets, I did not add aliases for them in `pkgs`, because I think they are (more) rarely used. But that could be done as well if desired.~~ <- changed my mind on that
Apart from these points I tried to be backwards compatible. Some of these backwards compat. devices can / should probably be deprecated and dropped eventually.

I took the liberty to rename the rt kernels from `-rt` to `_rt` (only in the `linuxKernel.kernels` attrset), because I think its more consistent like that, and we are breaking with old convention anyway. The aliases in `pkgs` remain unchanged, so as to not break existing configs. 

~~There is a separate commit, which~~ gets rid of the hack in `hardenedLinuxPackagesFor` as well.

###### Things done

* Compared the output of `lib.attrNames linuxPackages_5_12` on this branch and master, so that all kernel modules are still present. 
* Checked with `nix search .# linux_5_10` that `recurseForDerivations` is properly set, so kernel modules are found (and will be built by hydra).
 
<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

xaverdh committed:
top-level: move linux kernels, packages and related functions to linux-kernels.nix

teto commented:
I've wanted this for a long time ! ty

Atemu commented on pkgs/top-level/all-packages.nix:
```diff
@@ -20547,34 +20120,62 @@ in
-      modDirVersionArg = kernel.modDirVersion + (kernelPatches.hardened.${kernel.meta.branch}).extra;
-      isHardened = true;
-  });
+  # for backwards compatibility ---
+  # TODO: deprecate these?
```
Would it be possible to do all this automatically via a function?

xaverdh replied:
maybe with an overlay.. will look into that

xaverdh replied:
to clarify: the issue is that we have to merge attributes into `pkgs`, and I don't know how to do that in all-packages

xaverdh commented:
So I fixed the `makeExtensible` usage, but I don't know why the original code did not work (and why it worked as part of all-packages), which makes me sad. The issue was, that somehow `callPackage` was taken from the outer scope, but it should be the same in all-packages, since there is also an external `callPackage` in scope there..

xaverdh commented:
So I adjusted the old documentation. Still on my todo list is writing docs for the new functions.

xaverdh commented:
This should be in good state now.

> I kept the versioned linuxPackages_* and linux_* aliases for backwards compat., but I think they should either be considered deprecated in favor linuxKernel.packages.* and/or generated from these automatically (but I don't know how to that yet; help is very much appreciated here!).

That's the one outstanding to answer. Do we deprecate these, and keep the existing ones them around until they go EOL?
And somewhat orthogonal to that, is there a way to automatically inject them into `pkgs`?

xaverdh commented:
rebased

xaverdh commented:
rebased again
unsurprisingly, this does not age well..

xaverdh commented:
changed: better handling of `linux_latest_hardened`

xaverdh committed:
docs: adjust to the new structure of kernel packages

xaverdh committed:
nixos/doc: adjust to the new structure of kernel packages

Atemu commented on pkgs/top-level/all-packages.nix:
```diff
@@ -20154,490 +20154,107 @@ in
-
-    can-isotp = callPackage ../os-specific/linux/can-isotp { };
-  });
+  # Versioned aliases kept for backwards compatibility ------
+  # Don't add to this list!
+  linuxPackages_rt_5_4 = linuxKernel.packages.linux_rt_5_4;
+  linuxPackages_rt_5_10 = linuxKernel.packages.linux_rt_5_10;
+  linuxPackages_rt_5_11 = linuxKernel.packages.linux_rt_5_11;
+  linuxPackages_rpi1 = linuxKernel.packages.linux_rpi1;
+  linuxPackages_rpi2 = linuxKernel.packages.linux_rpi2;
+  linuxPackages_rpi3 = linuxKernel.packages.linux_rpi3;
+  linuxPackages_rpi4 = linuxKernel.packages.linux_rpi4;
+  linuxPackages_4_4 = linuxKernel.packages.linux_4_4;
+  linuxPackages_4_9 = linuxKernel.packages.linux_4_9;
+  linuxPackages_4_14 = linuxKernel.packages.linux_4_14;
+  linuxPackages_hardkernel_4_14 = linuxKernel.packages.hardkernel_4_14;
+  linuxPackages_4_19 = linuxKernel.packages.linux_4_19;
+  linuxPackages_5_4 = linuxKernel.packages.linux_5_4;
+  linuxPackages_5_10 = linuxKernel.packages.linux_5_10;
+  linuxPackages_5_11 = linuxKernel.packages.linux_5_11;
+  linuxPackages_5_12 = linuxKernel.packages.linux_5_12;
+
+  linux_mptcp_95 = linuxKernel.kernels.linux_mptcp_95;
+  linux_rpi1 = linuxKernel.kernels.linux_rpi1;
+  linux_rpi2 = linuxKernel.kernels.linux_rpi2;
+  linux_rpi3 = linuxKernel.kernels.linux_rpi3;
+  linux_rpi4 = linuxKernel.kernels.linux_rpi4;
+  linux_4_4 = linuxKernel.kernels.linux_4_4;
+  linux_4_9 = linuxKernel.kernels.linux_4_9;
+  linux_4_14 = linuxKernel.kernels.linux_4_14;
+  linux_4_19 = linuxKernel.kernels.linux_4_19;
+  linux_5_4 = linuxKernel.kernels.linux_5_4;
+  linux-rt_5_4 = linuxKernel.kernels.linux_rt_5_4;
+  linux_5_10 = linuxKernel.kernels.linux_5_10;
+  linux-rt_5_10 = linuxKernel.kernels.linux_rt_5_10;
+  linux_5_11 = linuxKernel.kernels.linux_5_11;
+  linux_5_12 = linuxKernel.kernels.linux_5_12;
+  linux-rt_5_11 = linuxKernel.kernels.linux_rt_5_11;
+  # ------
```
We need a way to communicate that these names are now legacy. How could we achieve this? I feel like a release note isn't going to reach most users. 

Not something that needs to happen in this PR but needs to be thought about at least.

xaverdh replied:
Yes I was hoping that most people use the unversioned packages anyway, and those that do pin versions usually keep an eye on what happens in nixpkgs.
This is a problem in general I think: we lack a reliable channel to communicate changes. Although the release notes are still the closest thing to that we have, I guess.
Maybe we can add traces to these at some point (saying that they are deprecated) while still not breaking old configs.

Atemu replied:
Clearly written eval warnings are definitely the best way of doing this kind of thing IMO.  
Maybe we could make these upcoming permanent renames a special kind of alias which prints a trace, similar to how nixos modues renames work.

I'll have a think about that tonight and open an issue tomorrow but feel free to go ahead and create one without me.

jonringer replied:
I don't think exposing the old `linuxPackages*` package sets is too bad. It would also avoid having users needing to update their configuration.nix

jonringer replied:
not sure what to do with the kernel variants

cc @samueldr

Atemu reviewed (approved):
Looks really good and thorough; great work!

xaverdh commented:
Note to self: If / once this is merged, rebase https://github.com/xaverdh/nixpkgs/commit/85e2a3809aff54970065d22d4aa23f57b1d4f03a on top, and open a follow up pr.

jonringer commented on pkgs/top-level/linux-kernels.nix:
```diff
@@ -0,0 +1,525 @@
+{ pkgs
```
I don't want to bike shed too much, but this should probably go in `pkgs/os-specific/linux`.

Then again, we have darwin packages. Probably relevant for a larger discussion.

cc @vcunat @Mic92 @NeQuissimus (and a bunch of others)

Mic92 replied:
I think `top-level` is a good place for package sets.

jonringer replied:
meh, I guess I don't feel strongly about it, and it seems that the current paradigm is to use `top-level`. I'll resolve this.

Mic92 reviewed (approved):
I have not reviewed all the code but I agree with the idea.

xaverdh commented:
changes:
* There is now a `packageAliases` set, which means one additional indirection for some attributes, but also ensures that all the version information (that needs to change) is contained in `linux-kernels.nix`. Also I think that it does capture what we are actually doing here.
* The libre kernels are now incorporated into the package set
* `linuxPackagesFor` is now just `packagesFor` in `linuxKernel` (still accessible under `pkgs.linuxPackagesFor` though)

xaverdh commented:
Maybe `linux_libre` should actually be a function?

jonringer commented:
I agree with the idea as well. But since this will be a pretty big shift. just going to cc a few others

cc @domenkozar @vcunat @edolstra @FRidh @NeQuissimus

jonringer commented:
Do you mind adding something in the release notes, under `notable changes`.

We may also want to add a `lib.warn` around using the older `linuxPackagesXXX`

jonringer commented:
please address conflicts

xaverdh committed:
nixos/rl-2111: mention changes to linux kernel infrastructure

xaverdh commented:
rebased, and added a release note entry.
Not sure about adding `lib.warn` just yet. I would like to have a shorthand function for `lib.warn "linuxKernel_foo was removed, use linuxKernel.packages.foo"`, but not sure where to put it (in `all-packages`?). Would be good to have something more general there.
 
It currently fails to eval anyway, because somehow the 5.11 kernel got erased in the master branch?

xaverdh commented:
Sorry, that was a bit messy (I am probably too tired right now). Should be good now I hope.
Given my confusion around 5.11, I just double checked that no kernels or modules got lost or added in the rebase, by comparing the tab completion of `linuxPackages_`  and the output of `lib.attrNames linuxPackages_5_12` in `nix repl` on this branch and master. Nonetheless if I have to rebase this again, whoever merges this pr might want to check again, just to be sure.

Atemu commented:
Migrating the attr names and warning users can be left to a separate PR IMO, this one already does a ton.

xaverdh commented:
rebased again, not sure how I managed to get the ofborg-internal-error thing..

jonringer commented:
rebase again, and ofborg should be re-triggered

cole-h commented:
@ofborg eval

jonringer commented:
there's merge conflicts, so it should be rebased regardless

xaverdh commented:
So what is required to get this merged?
I feel a bit tired of rebasing it every other day.

edolstra commented:
I'm not in favor of `latestPackageWithModules`. Nix has never tried to automatically find a configuration that satisfies some constraints for the user; instead the user has to specify the desired configuration. Finding a feasible configuration is inherently slow (and I dread the logical end status - somebody writing a SAT solver in Nixpkgs...).

xaverdh commented:
Well we kind of already do this by having various attributes like "linux", "linux_lastest" instead of the user just saying what precise version of the kernel they want. Although in principle there is a tradeoff between "convenience for the user" and "complexity of the implementation", I think in this case the implementation is simple enough not to be an issue.

Still if this is controversal, I could drop the respective commit for now. I think moving the packages into their own package set is valuable on its own (and its the part that causes merge conflicts). Would that be acceptable?

jonringer commented:
> Well we kind of already do this by having various attributes like "linux", "linux_lastest" instead of the user just saying what precise version of the kernel they want. Although in principle there is a tradeoff between "convenience for the user" and "complexity of the implementation", I think in this case the implementation is simple enough not to be an issue.
> 
> Still if this is controversal, I could drop the respective commit for now. I think moving the packages into their own package set is valuable on its own (and its the part that causes merge conflicts). Would that be acceptable?

We can move it to a different PR. This is currently tackling more than one aspect of linux-packages

xaverdh commented:
I changed this pr to only move the linux package set (as its title already claims).
Will open a separate pr for the new functionality, then we can discuss this further.

jonringer commented:
this looks to be a 1000+ rebuild (not sure why), please target staging or eliminate the cause of the rebuild.

xaverdh commented:
> this looks to be a 1000+ rebuild (not sure why), please target staging or eliminate the cause of the rebuild.

I am not sure why, actually this should not cause any rebuilds. Still I can open a new pr targeting staging.

xaverdh commented:
closing in favour of #128134 (which is targeted at staging)
